src/NHibernate/Tuple/Component/PocoComponentTuplizer.cs

```diff
@@ -52,6 +54,7 @@ public PocoComponentTuplizer(Mapping.Component component)
 				parentGetter = parentProperty.GetGetter(componentClass);
 			}
 
+			isBytecodeProviderImpl = Cfg.Environment.BytecodeProvider is BytecodeProviderImpl;
```

Comment:
It should be moved into `SetReflectionOptimizer`, because nothing guarantees the byte code provider will be the same on deserialization. By the way the test could be on the `optimizer.AccessOptimizer` concrete implementation instead, which would be more straightforward.